### PR TITLE
[FLINK-1752] Improvements to Kafka Source/Sink & some bugfixes in Flink Streaming

### PR DESCRIPTION
**Pull request by rmetzger**
Hi,
I've spend the last two days on improving the Kafka connectors and the streaming fault tolerance.
Since its the first time I'm touching this code, I would like to get some feedback from the streaming guys (in particular on the changes with the state handling).

Please note that this PR is based on https://github.com/apache/flink/pull/474, so it also contains commits from @gaborhermann.
I'll squash the commits after the first round of feedback.

If you want I can separate the the critical fixes from the Kafka connector rewrite, but i would prefer merging everything at once.

### CODE REVIEW

File: flink-java/src/main/java/org/apache/flink/api/java/DataSet.java [hunk 1]
```diff
@@ -182,7 +182,6 @@ protected void fillInType(TypeInformation<T> typeInfo) {
 		if (getExecutionEnvironment().getConfig().isClosureCleanerEnabled()) {
 			ClosureCleaner.clean(f, true);
 		}
-		ClosureCleaner.ensureSerializable(f);
 		return f;
```
Comment (mbalassi) [inline flink-java/src/main/java/org/apache/flink/api/java/DataSet.java]: Why remove this?

File: flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/StreamCheckpointCoordinator.scala [hunk 2]
```diff
@@ -75,17 +75,17 @@ class StreamCheckpointCoordinator(val executionGraph: ExecutionGraph,
     case InitBarrierScheduler =>
       context.system.scheduler.schedule(interval,interval,self,BarrierTimeout)
       context.system.scheduler.schedule(2 * interval,2 * interval,self,CompactAndUpdate)
-      log.debug("[FT-MONITOR] Started Stream State Monitor for job {}{}",
+      log.info("Started Stream State Monitor for job {}{}",
         executionGraph.getJobID,executionGraph.getJobName)
```
Comment (mbalassi) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/StreamCheckpointCoordinator.scala]: +1, this is definitely info level.

File: flink-staging/flink-streaming/flink-streaming-connectors/pom.xml [hunk 3]
```diff
@@ -46,9 +53,13 @@ under the License.
 		<dependency>
 			<groupId>org.apache.kafka</groupId>
 			<artifactId>kafka_2.10</artifactId>
-			<version>0.8.2.0</version>
+			<version>${kafka.version}</version>
 			<exclusions>
```
Comment (mbalassi) [inline flink-staging/flink-streaming/flink-streaming-connectors/pom.xml]: Heads-up: The code needed for Kafka 0.8.2.0 is slightly different.

File: flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/simple/SimpleKafkaSource.java [hunk 4]
```diff
@@ -71,7 +67,7 @@ public void cancel() {
 
 	@Override
 	public void open(Configuration config) throws InterruptedException {
-		initializeConnection();
+//		initializeConnection();
```
Comment (mbalassi) [inline flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/simple/SimpleKafkaSource.java]: Why is this commented out? Eventually we should delete the Kafka connectors using the simple API, there is not much point to keeping them at the current point.

File: flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/simple/iterator/KafkaOnePartitionIterator.java [hunk 5]
```diff
@@ -41,22 +43,25 @@
 /**
  * Iterates the records received from a partition of a Kafka topic as byte arrays.
  */
-public class KafkaConsumerIterator implements Serializable {
+public class KafkaOnePartitionIterator implements KafkaConsumerIterator, Serializable {
```
Comment (mbalassi) [inline flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/simple/iterator/KafkaOnePartitionIterator.java]: I'd prefer KafkaSinglePartitionIterator

File: flink-staging/flink-streaming/flink-streaming-connectors/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaITCase.java [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+
+public class KafkaITCase {
+
+	private static final Logger LOG = LoggerFactory.getLogger(KafkaITCase.class);
+
+	private final String TOPIC = "myTopic";
+
+	private final int ZK_PORT = 6667;
+	private final int KAFKA_PORT = 6668;
+	private String kafkaHost;
```
Comment (mbalassi) [inline flink-staging/flink-streaming/flink-streaming-connectors/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaITCase.java]: These ports maybe already allocated, I'd use NetUtils to get an available one.

File: flink-java/src/main/java/org/apache/flink/api/java/DataSet.java (hunk 1 above)
Comment (rmetzger) [inline flink-java/src/main/java/org/apache/flink/api/java/DataSet.java]: Oh, you're right. I'll add it again.

File: flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/simple/iterator/KafkaOnePartitionIterator.java (hunk 5 above)
Comment (gaborhermann) [inline flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/simple/iterator/KafkaOnePartitionIterator.java]: +1, changing it.

File: flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/simple/SimpleKafkaSource.java (hunk 4 above)
Comment (gaborhermann) [inline flink-staging/flink-streaming/flink-streaming-connectors/src/main/java/org/apache/flink/streaming/connectors/kafka/api/simple/SimpleKafkaSource.java]: I agree, no point in keeping SimpleKafkaSource, deleted it.

File: flink-staging/flink-streaming/flink-streaming-connectors/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaITCase.java (hunk 6 above)
Comment (gaborhermann) [inline flink-staging/flink-streaming/flink-streaming-connectors/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaITCase.java]: Thanks for the hint, using it.

File: flink-staging/flink-streaming/flink-streaming-connectors/pom.xml (hunk 3 above)
Comment (rmetzger) [inline flink-staging/flink-streaming/flink-streaming-connectors/pom.xml]: Mh, thats not so easy to solve. We probably have to create different maven modules for the different supported versions (if its not possible to detect and use different methods at runtime)

Comment (rmetzger) [inline flink-staging/flink-streaming/flink-streaming-connectors/pom.xml]: If this is only about the different names of `leader.getConnectionString()`, we can also play some reflection tricks.

**Comment by rmetzger** [comment]
We're addressed all comments on the pull request.

I'm currently running some tests: https://travis-ci.org/rmetzger/flink/builds/55061414.
I'll merge the change once they've passed.

**Pull request closed without merging**